### PR TITLE
Monitor runs from server

### PR DESCRIPTION
**Pull request by Hedingber**
Don't be horified by the number of lines changed it's mostly test code moved from one place to another.

**Terminology for this description:**
* Monitoring - keeping the Run state updated.
* Log collection - Collecting the logs of a Run and saving them in MLRun's persistency.

**What we had before this PR:**
**Monitoring:**
There was basically 2 mechanisms:
* Client - the MLRun SDK code running inside the user's Jobs (wrapping its code) updated the Run state according to progress - e.g. to `running` when it started/to `completed` when it finished/to `error` when an error was raised
* Client + Server - When a user was running a job using the SDK (calling `.run()`) by default `watch=True` which means that the SDK will poll the API's `GET /log/{project}/{uid}` endpoint to get the run logs. In the API's code of this endpoint there was a hack - since it anyways gets the Job's pods, it looked on their state, and updated the Run's state accordingly.
So practically get logs endpoints was abused to monitor Run status, and Client log polling was abused to trigger the monitoring periodically.

**Log collection:**
The get logs endpoint has two possible sources, MLRun's persistency and K8s, when there is something in the persistency it reads from there, otherwise it fall back to K8s. Practically nothing was collecting the logs from K8s to persistency, so when clients did get logs, the API basically just proxied them to K8s.
The exception is the cleanup mechanism, which before it cleans up job's resources, it collects the logs and saves them to persistency, so eventually the logs were getting to the persistency.

**What this PR does:**
**Monitoring:**
* Client - didn't touch, this will keep working the same way, I think that the SDK running in the Job has the most accurate information about the real state, and therefore it should continue being part of the monitoring mechanism, also since it's working in kind of PUSHing way (and not polling), it gives almost zero latency between real state and the time the Run state is updated.
* Server - removed any logic to update Run state from get logs endpoint, this endpoint now do what it should - give logs.
Added monitoring logic that runs periodically (by default every 5 seconds, configurable) in background and updates the Runs states.

**Log collection:**
Not much changed, I think that trying to move logs from K8s to our persistency while the job is running is a waste of time, either we'll poll K8s too often and do a lot of un-needed pressure on it, or we'll poll it too rarely so the logs in our persistency (which will be the ones served to the user) won't be updated/relevant.
The only thing I did change, is that when the monitoring identifies a run reached stable state (completed/failed) it collects the logs from K8s and push to our persistency.

Other changes:
* Changed pipelines job to use the API to run jobs (otherwise it won't monitor them) instead of running them indepndently.
* Moved runtime handlers tests from `tests/runtimes/test_runtime_handlers` to `tests/api/runtime_handlers` and splitted them to test file per handler, did it cause I needed some of the API's testing facilities.
Note - those handlers are running only in context of the API, therefore the right thing to do is to also move the runtime handlers themselves out of the `mlrun/runtimes` package to somewhere under `mlrun/api` as part of the effort to distinct SDK code from API code. this PR is already too big so I didn't do it here, will probably do it in some followup PR.

### CODE REVIEW

File: mlrun/runtimes/base.py [hunk 1]
```diff
@@ -811,6 +811,52 @@ def delete_runtime_object_resources(
...
+    ):
+        logger.debug(
+            "Starting run monitor loop",
+            project=project,
+            run_uid=run_uid,
+            timeout=timeout,
+            interval=interval,
+        )
+        try:
+            mlrun.utils.helpers.retry_until_successful(
```
Comment (omesser) [inline mlrun/runtimes/base.py:831]: I know it's good for now (it works), but I think the `retry_until_successful` pattern is not fitting to this kind of monitoring loop. Also I saw mlrun's retry_until_successful is synchronous.
This should be something like this: https://stackoverflow.com/questions/37512182/how-can-i-periodically-execute-a-function-with-asyncio (even though you're in a threadpool)

File: mlrun/runtimes/base.py [hunk 2]
```diff
@@ -1086,11 +1131,15 @@ def _pre_deletion_runtime_resource_run_actions(
             )
             raise ValueError("Could not resolve run uid from runtime resource")
 
-        self._ensure_runtime_resource_run_status_updated(
-            db, db_session, project, uid, desired_run_state
+        logger.info(
+            "About to cleanup runtime resources, pre-deletion actions will be performed",
```
Comment (omesser) [inline mlrun/runtimes/base.py:1135]:
```suggestion
            "Cleaning up runtime resources and running pre-deletion actions",
```
"About to" / "Starting to" ... same. needless fluff
Also are you sure this log is in the right place? this function doesn't have any calls to any "cleanup" or "pre-deletion" functions after that. just `_ensure_run_status_updated` and `_ensure_finished_run_logs_collected` which sounds very different (maybe just a naming issue)

File: mlrun/runtimes/base.py [hunk 3]
```diff
@@ -1121,10 +1170,84 @@ def _is_runtime_resource_run_in_transient_state(
...
+            db, db_session, project, run_uid
+        )
+        updated_run_state = self._ensure_run_status_updated(
+            db, db_session, project, run_uid, desired_run_state
+        )
+
+        if updated_run_state not in RunStates.stable_states():
+            raise Exception("run did not reach stable state yet")
+
+    def _resolve_run_state_representing_current_resource_state(
```
Comment (omesser) [inline mlrun/runtimes/base.py:1189]:
```suggestion
    def _resolve_run_state(
```
🤷  am i missing something?

File: mlrun/runtimes/base.py [hunk 4]
```diff
@@ -1121,10 +1170,84 @@ def _is_runtime_resource_run_in_transient_state(
 
         return False, last_update
 
+    def _verify_run_reached_stable_state_and_monitor_progress(
+        self, db: DBInterface, db_session: Session, project: str, run_uid: str
+    ):
+        """
+        This function runs in a retry_until_successful and therefore need to raise if run didn't reach stable state
+        """
+        desired_run_state = self._resolve_run_state_representing_current_resource_state(
```
Comment (omesser) [inline mlrun/runtimes/base.py:1179]: Maybe this is better cleared offline but you're using the word "desired" here on the object that represents the REAL state.
Usually desired_state (== target_state) is a different concept.
This is also true inside the helper functions `desired_run_state` is used in a way that I think is not accurate

File: mlrun/runtimes/base.py [hunk 5]
```diff
@@ -1121,10 +1170,84 @@ def _is_runtime_resource_run_in_transient_state(
...
+            crd_version,
+            namespace,
+            crd_plural,
+            label_selector=label_selector,
+        )
+        if not crd_objects or len(crd_objects["items"]) == 0:
+            raise RuntimeError("Run crd object could not be found")
+        if len(crd_objects["items"]) > 1:
+            logger.warning(
+                "Unexpectedly received more than one crd object for run. Best effort - using the first one"
```
Comment (omesser) [inline mlrun/runtimes/base.py:1222]:
```suggestion
                "Received more than one crd record for run. Heuristically using the first one"
```
Can this happen? should this not be a real severe error?

File: mlrun/runtimes/base.py [hunk 6]
```diff
@@ -1148,42 +1271,58 @@ def _ensure_runtime_resource_run_logs_collected(
...
+                uid=uid,
+                desired_run_state=desired_run_state,
+            )
+        else:
+            current_run_state = run.get("status", {}).get("state")
+            if current_run_state:
+                if current_run_state == desired_run_state:
+                    update_run = False
+                # if the current run state is stable and different then the desired - don't touch
+                if current_run_state in RunStates.stable_states():
```
Comment (omesser) [inline mlrun/runtimes/base.py:1307]: The comment repeats what the code does not EXPLAINS why.
This seems, at least to the uninformed reader, wrong - the `desired_run_state` is actually extracted from the actual k8s object (crd / pod) - so this is the "real" or "current" state.
The `current_run_state` is coming from mlrun's persistency, and is expected to be out of sync.
We're allegedly here updating mlrun's db state as per the real status. So why does mlrun's stored state trumps the real one?
The method's name says it "ensures" it's updated, here you choose to not update it.

File: mlrun/runtimes/base.py [hunk 7]
```diff
@@ -1148,42 +1271,58 @@ def _ensure_runtime_resource_run_logs_collected(
...
                 db_session, project, uid, source=LogSources.K8S
             )
-            crud.Logs.store_log(logs_from_k8s, project, uid, append=False)
+            if logs_from_k8s:
+                logger.info("Storing finished run logs", project=project, uid=uid)
+                crud.Logs.store_log(logs_from_k8s, project, uid, append=False)
 
     @staticmethod
-    def _ensure_runtime_resource_run_status_updated(
+    def _ensure_run_status_updated(
```
Comment (omesser) [inline mlrun/runtimes/base.py:1282]:
```suggestion
    def _update_run_state(
```
Also, isn't there some "Run" class that you can implement this as  `run.update_state(...)` on ?

File: mlrun/runtimes/constants.py [hunk 8]
```diff
@@ -39,14 +55,118 @@ def default():
...
+            RunStates.completed,
+            RunStates.error,
+            RunStates.running,
+            RunStates.created,
+            RunStates.pending,
+            RunStates.unknown,
+        ]
 
     @staticmethod
     def stable_states():
```
Comment (omesser) [inline mlrun/runtimes/constants.py:75]: wrt our past discussion - "stable" states are not the right term here in my opinion.
these are not representing stateful resources like "cluster" "tenant" or "user" - but runtime resources. Those resources spend most of their time "running" until they complete/fail.
So it's a bit funny referring to the "final/terminal" states as stable, which insinuates those resources are "unstable" during their designated runtime even though "running" is what they do 99% of the time :)

File: mlrun/runtimes/sparkjob.py [hunk 9]
```diff
@@ -340,19 +341,17 @@ def _resolve_crd_object_status_info(
         # it is less likely that there will be new stable states, or the existing ones will change so better to resolve
         # whether it's a transient state by checking if it's not a stable state
         state = crd_object.get("status", {}).get("applicationState", {}).get("state")
-        in_transient_state = state not in ["COMPLETED", "FAILED"]
-        desired_run_state = None
+        in_transient_state = state not in SparkApplicationStates.stable_states()
```
Comment (omesser) [inline mlrun/runtimes/sparkjob.py:344]:
```suggestion
        in_progress = state not in SparkApplicationStates.stable_states()
```

File: tests/rundb/test_httpdb.py [hunk 10]
```diff
@@ -132,20 +132,30 @@ def cleanup():
...
         return Server(url, conn, workdir)
 
     def cleanup():
-        if proc:
-            proc.terminate()
+        if process:
+            process.terminate()
+            stdout = process.stdout.read()
+            human_readable_stdout = (
+                stdout.decode("utf-8").replace("\\n", "\n").replace("\\t", "\t")
```
Comment (omesser) [inline tests/rundb/test_httpdb.py:151]: this: https://stackoverflow.com/a/24519338/1875031
or this: https://stackoverflow.com/a/37059682/1875031 (I'd go with this one personally)

**Review by omesser — CHANGES_REQUESTED** [review]

**Comment by Hedingber** [comment]
Commenting so I won't forget, discussed with @yaronha offline about the implementation, he suggested that instead of running a monitor loop (which query k8s every X seconds) **per job**, it would make much more sense to just run one loop, always, which will monitor all jobs runs

File: mlrun/api/api/utils.py [hunk 11]
```diff
@@ -123,12 +128,61 @@ def _parse_submit_job_body(db_session: Session, data):
     return function, task
 
 
-def submit(db_session: Session, data):
+async def submit_job(db_session: Session, data):
+    project, function_kind, run_uid, response = await run_in_threadpool(
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/api/api/utils.py:132]: Run it is

[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/api/api/utils.py:132]: Actually for this function specifically I think the right terminology is `submit_task` cause you're actually submitting a task to run on a function, and the outcome is a Run

File: mlrun/api/api/utils.py [hunk 12]
```diff
@@ -165,7 +222,5 @@ def submit(db_session: Session, data):
             HTTPStatus.BAD_REQUEST.value, reason="runtime error: {}".format(err)
         )
 
-    logger.info("response: %s", response)
-    return {
-        "data": response,
-    }
+    logger.info("Submit job succeeded, returning response", response=response)
+    return project, fn.kind, run_uid, {"data": response}
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/api/api/utils.py:226]: Not really this is just the way it is

File: mlrun/api/crud/logs.py [hunk 13]
```diff
@@ -26,49 +28,32 @@ def get_log(
...
+            1. str of the run state (so watchers will know whether to continue polling for logs)
+            2. str of the log itself
+        """
         out = b""
         log_file = log_path(project, uid)
-        pod_status = None
+        data = get_db().read_run(db_session, uid, project)
+        if not data:
+            log_and_raise(HTTPStatus.NOT_FOUND.value, project=project, uid=uid)
+        run_state = get_in(data, "status.state", "")
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/api/crud/logs.py:42]: yes best effort, it is `get_in` cause that's just how it was, changed to `data.get('status', {}).get('state', '')`

File: mlrun/api/db/sqldb/db.py [hunk 14]
```diff
@@ -94,6 +94,9 @@ def _delete_logs(self, session: Session, project: str):
 
     def store_run(self, session, struct, uid, project="", iter=0):
         project = project or config.default_project
+        logger.debug(
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/api/db/sqldb/db.py:97]: Changed to `run_data`

File: mlrun/api/main.py [hunk 15]
```diff
@@ -140,6 +142,53 @@ def _start_periodic_cleanup():
...
+def _resume_monitoring_runs():
+    logger.info("Resuming monitoring runs")
+    db_session = create_session()
+    try:
+        projects = get_db().list_projects(db_session)
+        for project in projects:
+            runs = get_db().list_runs(db_session, project=project.name)
+            for run in runs:
+                try:
+                    state = run.get("status", {}).get("state")
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/api/main.py:154]: I prefer it too, where-ever I touch/ add code I try to move using it

File: mlrun/api/main.py [hunk 16]
```diff
@@ -140,6 +142,53 @@ def _start_periodic_cleanup():
     run_function_periodically(int(config.runtimes_cleanup_interval), _cleanup_runtimes)
 
 
+def _resume_monitoring_runs():
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/api/main.py:145]: It starts monitoring only for Runs not in stable state (to cover the case where the API got restarted when something was running and monitored)
Anyways as I said in my comment, launching a monitor loop **per run** seems to be not a good idea, going to change it to one loop that will just run always so this whole thing will be removed

File: mlrun/api/utils/singletons/k8s.py [hunk 17]
```diff
@@ -1,17 +1,5 @@
...
-    return k8s
-
-
-def initialize_k8s():
-    global k8s
-    try:
-        k8s = K8sHelper()
-    except Exception:
-        pass
+    return get_k8s_helper(silent=True)
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/api/utils/singletons/k8s.py:5]: Here we are under the `api` module, and this is the module for `singletons` for the API.
the `k8s_utils.py` which `get_k8s_helper()` comes from, and as you said already manages a singleton, is used both by the API and the SDK and therefore can't be here.
When I wrote this file `k8s.py` I didn't notice that there are actually double management of singletons, only now I did, so fixed it in this PR, and made this thing here just a proxy to there.
I can basically remove the proxy and reference the original from everywhere, but It will make tons of changes in a lot of files so I prefer not to, also I think that maybe eventually there will be something here designeted only for the API usage so another reason to currently keep it

File: mlrun/runtimes/base.py (hunk 1 above)
Comment (Hedingber) [reply mlrun/runtimes/base.py:831]: I'm not sure what you mean by "not fitting to this kind of monitoring loop", why ?
exactly because I'm in a threadpool I can not use asyncio or do async stuff

File: mlrun/runtimes/base.py (hunk 2 above)
Comment (Hedingber) [reply mlrun/runtimes/base.py:1135]: This log is inside the pre-deletion function, and the pre-deletion actions are exactly these 2, ensuring run status and ensuring logs collected.
changed to `"Performing pre-deletion actions before cleaning up runtime resources"`

File: mlrun/runtimes/base.py (hunk 3 above)
Comment (Hedingber) [reply mlrun/runtimes/base.py:1189]: I would expect function like `_resolve_run_state` to go to the DB, query for the run, find the state and return it.
While this function is actually checking the runtime resource, resolving its state, and translating it to what should be the representing run state

File: mlrun/runtimes/base.py (hunk 4 above)
Comment (Hedingber) [reply mlrun/runtimes/base.py:1179]: Explained above the idea of this function
> This function is actually checking the runtime resource, resolving its state, and translating it to what should be the representing run state
I agree both function and variable naming are not great, I accept suggestions if you have something

File: mlrun/runtimes/base.py (hunk 5 above)
Comment (Hedingber) [reply mlrun/runtimes/base.py:1222]: Not really, I mean, if this happens it means that two CRD objects somehow got the `mlrun/uid` label with the same value, which is very weird, question is what should monitoring do ? just give up and leave this run to whatever the state currently is ?

File: mlrun/runtimes/base.py (hunk 6 above)
Comment (Hedingber) [reply mlrun/runtimes/base.py:1307]: As I explained in the PR description, beside of this logic, the SDK running inside the Run pod itself (wrapping the user's code) also updates the Run's state.
So the scenario I was thinking about when doing this is:
The user's code finishes successfully, therefore the SDK updating the Run's state to `Completed` now for some reason whether because some bug in the wrapping SDK code, or I don't know, suddenly OOM kill to the Pod, the Pod got into Failed state.
So although in this scenario the user's code executed successfully we want to mark the Run as failed ? 
I'm not sure, it's a weird state, both decisions seems legit, anyways we must identify this state and log it.
what do you think ?

File: mlrun/runtimes/base.py (hunk 7 above)
Comment (Hedingber) [reply mlrun/runtimes/base.py:1282]: The Run class is for the SDK, no such a thing for the API

File: mlrun/runtimes/constants.py (hunk 8 above)
Comment (Hedingber) [reply mlrun/runtimes/constants.py:75]: so what is your suggestion ? terminal states ? if so, what's the opposite, non-terminal ? transient ? running ?

Comment (Hedingber) [reply mlrun/runtimes/constants.py:75]: saw you suggested `in_progress` instead of `in_transient_states` that seems a bit weird cause a Pod in pending state is not exactly in progress

File: mlrun/runtimes/sparkjob.py (hunk 9 above)
Comment (Hedingber) [reply mlrun/runtimes/sparkjob.py:344]: move the discussion on this to previous thread

File: mlrun/runtimes/base.py (hunk 5 above)
Comment (Hedingber) [reply mlrun/runtimes/base.py]: This function is not used anymore (as part of moving to one loop to monitor all runs), but since I think it may be useful in the future I kept it

File: mlrun/runtimes/base.py (hunk 4 above)
Comment (Hedingber) [reply mlrun/runtimes/base.py]: This function does not exist anymore (as part of moving to one loop to monitor all runs)
but `desired_run_state` still exists and is the output of either `_resolve_pod_status_info` or `_resolve_crd_object_status_info`

File: mlrun/runtimes/base.py (hunk 3 above)
Comment (Hedingber) [reply mlrun/runtimes/base.py]: This function does not exist anymore (as part of moving to one loop to monitor all runs)

File: mlrun/runtimes/base.py (hunk 1 above)
Comment (Hedingber) [reply mlrun/runtimes/base.py]: This call does not exist anymore (as part of moving to one loop to monitor all runs), it's now just 
`run_function_periodically(int(config.runs_monitoring_interval), _monitor_runs)` which I think addresses your comment

File: mlrun/api/api/utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (omesser) [reply mlrun/api/api/utils.py]: 😆

File: mlrun/runtimes/base.py (hunk 3 above)
Comment (omesser) [reply mlrun/runtimes/base.py]: `_resolve_current_run_state` would've addressed you first comment (what you describe sounds more like a "get" or "fetch" to me)
But, like you wrote, not relevant now I suppose

File: tests/conftest.py [hunk 18]
```diff
@@ -116,3 +118,49 @@ def init_sqldb(dsn):
     engine = create_engine(dsn)
     Base.metadata.create_all(bind=engine)
     return sessionmaker(bind=engine)
+
+
+class DictToObjectWrapper:
```
Comment (omesser) [inline tests/conftest.py:123]: https://github.com/mewwts/addict or
https://github.com/Infinidat/munch - an actively maintained munch fork by the guys at https://github.com/Infinidat) or
https://github.com/ramazanpolat/prodict - this is a smaller one, but the author promises auto-completion in IDEs which sounds fantastic. I don't know if that works with the others, but if not, then this is a big advantage

I suggest you read this SO answer if you haven't come across it already (https://stackoverflow.com/a/59379520/1875031), or the entire thread if you're bored :)

File: tests/conftest.py [hunk 19]
```diff
@@ -116,3 +118,49 @@ def init_sqldb(dsn):
...
+                )
+            else:
+                setattr(
+                    self,
+                    key,
+                    DictToObjectWrapper(value) if isinstance(value, dict) else value,
+                )
+
+
+class DictToK8sObjectWrapper(DictToObjectWrapper):
```
Comment (omesser) [inline tests/conftest.py:156]: ?

**Review by omesser — CHANGES_REQUESTED** [review]

File: mlrun/api/api/utils.py [hunk 20]
```diff
@@ -123,12 +125,26 @@ def _parse_submit_job_body(db_session: Session, data):
     return function, task
 
 
-def submit(db_session: Session, data):
+async def submit_task(db_session: Session, data):
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/api/api/utils.py:128]: `_parse_submit_run_body` was just a left over from the first renaming (job -> run, after which I thought that task is more appropriate)...
as I said in my first comment, to me it sounds like you're submitting a Task (to be executed by the function) and the output is a Run object, you're not submitting a Run. (i.e. since the verb is submit, it feels like the noun should be what you give not what you get)
here is exactly the place where MLRun "gets a task" - the data here has a task.
Looking on the docs of MLRun's basic components - https://github.com/mlrun/mlrun/#basic-components:
> The run object is created as a result of running a task on a function
(IMO the verb should have been submitting instead running)
WDYT ?

File: mlrun/api/api/utils.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/api/api/utils.py:188]: needed for BC for the API

File: mlrun/runtimes/constants.py (hunk 8 above)
Comment (Hedingber) [reply mlrun/runtimes/constants.py:75]: Convinced 👍 , changing to terminal and non-terminal

File: mlrun/runtimes/base.py [hunk 21]
```diff
@@ -842,32 +877,31 @@ def _resolve_crd_object_status_info(
...
-        in_transient_state = pod.status.phase not in PodPhases.stable_phases()
-        desired_run_state = None
+        in_transient_state = pod["status"]["phase"] not in PodPhases.stable_phases()
         completion_time = None
+        desired_run_state = PodPhases.pod_phase_to_run_state(pod["status"]["phase"])
         if not in_transient_state:
-            desired_run_state = PodPhases.pod_phase_to_run_state(pod.status.phase)
-            for container_status in pod.status.container_statuses:
-                if hasattr(container_status.state, "terminated"):
+            for container_status in pod["status"].get("container_statuses", []):
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/runtimes/base.py:899]: It was defnied before this line, renamed

File: mlrun/runtimes/base.py [hunk 22]
```diff
@@ -1086,11 +1123,15 @@ def _pre_deletion_runtime_resource_run_actions(
...
+        logger.info(
+            "Performing pre-deletion actions before cleaning up runtime resources",
+            project=project,
+            uid=uid,
         )
 
-        self._ensure_runtime_resource_run_logs_collected(db, db_session, project, uid)
+        self._ensure_run_state(db, db_session, project, uid, desired_run_state)
+
+        self._ensure_finished_run_logs_collected(db, db_session, project, uid)
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/runtimes/base.py:1134]: same answer as - https://github.com/mlrun/mlrun/pull/467#discussion_r502182303
the log above says: `Performing pre-deletion actions before cleaning up runtime resources`, the code you're looking at is inside the function `_pre_deletion_runtime_resource_run_actions`, this function is called before the cleanup step which is happening after the call to this function, in the calling code

File: mlrun/runtimes/base.py [hunk 23]
```diff
@@ -1121,10 +1162,132 @@ def _is_runtime_resource_run_in_transient_state(
...
+        return f"mlrun/project={project},mlrun/uid={run_uid}"
+
     @staticmethod
-    def _ensure_runtime_resource_run_logs_collected(
+    def _ensure_finished_run_logs_collected(
         db: DBInterface, db_session: Session, project: str, uid: str
     ):
+        """
+        NOTE: This function logic based on the fact that the runtime resource was verified to reach stable state
+        therefore the naming FINISHED_runtime_resource
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/runtimes/base.py:1289]: Actually most of the logic here (regarding the mtime checks) is not relevant anymore, removed it, and renamed to `_collect_run_logs`

File: mlrun/runtimes/base.py [hunk 24]
```diff
@@ -1148,42 +1311,62 @@ def _ensure_runtime_resource_run_logs_collected(
...
+                crud.Logs.store_log(logs_from_k8s, project, uid, append=False)
 
     @staticmethod
-    def _ensure_runtime_resource_run_status_updated(
+    def _ensure_run_state(
         db: DBInterface,
         db_session: Session,
         project: str,
         uid: str,
         desired_run_state: str,
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/runtimes/base.py:1327]: I explained here - https://github.com/mlrun/mlrun/pull/467#discussion_r502185292 why there might be two different terminal states
anyways:
1. changed to `run_state` and `db_run_state`
2. changed so that `run_state` will take precedence over `db_run_state` always

File: mlrun/runtimes/constants.py [hunk 25]
```diff
@@ -39,14 +55,118 @@ def default():
...
+            SparkApplicationStates.submission_failed: RunStates.running,
+            SparkApplicationStates.pending_rerun: RunStates.running,
+            SparkApplicationStates.invalidating: RunStates.running,
+            SparkApplicationStates.succeeding: RunStates.running,
+            SparkApplicationStates.failing: RunStates.running,
+            SparkApplicationStates.unknown: RunStates.unknown,
+        }[spark_application_state]
+
+
+class MPIJobV1Alpha1States:
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/runtimes/constants.py:140]: I remember I started from this and It led to import cycles, IMO better to keep them all here

File: tests/conftest.py (hunk 18 above)
Comment (Hedingber) [reply tests/conftest.py:123]: I've read it, and took my solution from some other thread that is referenced somewhere in this thread (can't find it currently)
Generally, all of these packages gives the option to have dict which can be **also** access with object style (`obj.key`), **but** the dict functionalities remain (`.items()`, `.keys()` etc...) + dict style access is still possible (`obj['key']`)
I don't need these, therefore my solution can be much more simpler, all I need is a way to instantiate a mock object from a dict. 
I don't need to change anything on the dict afterwards, I don't need them to be synced, I don't need the object to keep act like a dict, I actually do no want it, e.g. if someone is doing `obj.keys()` is should raise an exception.
Specifically - prodict gives auto-completion if you define (meaning you know it) the schema from ahead, this is not our case, if it was, I could simply use the already in use Pydantic

File: tests/conftest.py (hunk 19 above)
Comment (Hedingber) [reply tests/conftest.py:156]: This is extending the generalized `DictToObjectWrapper` with functionality only mocked k8s objects should have (like `to_dict()` method)

File: mlrun/api/api/utils.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (omesser) [reply mlrun/api/api/utils.py:128]: Again - in plain english I would agree, and it makes sense considering that sentence, except the sentence makes no sense since the docs make a mishmash of what things are and says "running" a task.
But as for the mlrun definition - I think this is slightly inaccurate. It only makes sense because input is improperly named "task".
So it makes "english" sense (you submit a task, but you say that because a task is a long running process that submitting it means launching it). But, imagine that we rename "task" to what it is - "run_input" or "inputs"
now it's not reflecting, I think, what you want to reflect - "submit_run_inputs". I *think* the naming here tries to reflect that the "run" is launched, not that the "inputs" are ingested.

File: mlrun/runtimes/base.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (omesser) [reply mlrun/runtimes/base.py:900]: I meant 1 line before :)
But no functional different just defining it at the scope it's needed

File: mlrun/runtimes/base.py (hunk 2 above)
Comment (omesser) [reply mlrun/runtimes/base.py]: So log referring to cleanup should move to a function aware of the cleanup. It's not a huge sin but it's nicer that the logs don't exceed the scope of the logic. it's a better practice as you an move code around without having to change all the log lines because they all refer to where they are in the "grand scope" of things which, abstraction-wise, they shouldn't know about

File: mlrun/runtimes/base.py (hunk 6 above)
Comment (omesser) [reply mlrun/runtimes/base.py]: i think a completed job/pod cannot fail after it's completed - completed is a state k8s puts after the pod's containers were terminated. once a runtime resource is marked completed, that's supposed to be it. so can't OOM kill a non running container

File: mlrun/runtimes/base.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (omesser) [reply mlrun/runtimes/base.py]: Answered in original thread: https://github.com/mlrun/mlrun/pull/467#discussion_r504944887 resolving this thread to reduce noise

File: mlrun/runtimes/base.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (omesser) [reply mlrun/runtimes/base.py]: 😱

File: mlrun/runtimes/base.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (omesser) [reply mlrun/runtimes/base.py]: Answered in original thread: https://github.com/mlrun/mlrun/pull/467#discussion_r504947101
Resolving this one to reduce noise

File: tests/conftest.py (hunk 18 above)
Comment (omesser) [reply tests/conftest.py:123]: I would strongly suggest to go with one of these (that fits your needs) even if you just need a subset of the functionality instead of re-inventing the wheel (there's a huge selection of premium wheels - pick one that suits you) - but, 🤷  as you wish

File: tests/conftest.py (hunk 19 above)
Comment (omesser) [reply tests/conftest.py:156]: Neither this nor `DictToObjectWrapper` mention "mock" or inherit from Mock object.
If they are meant to be used for mocking they should inherit from Mock and named as such I suppose 🤷

File: tests/conftest.py (hunk 18 above)
Comment (Hedingber) [reply tests/conftest.py:123]: I don't want to use them not because of the extra functionality I don't need, but because the extra functionality I don't want, e.g. if somewhere in the code the runtime variable is an object, and someone mistakenly do `obj['key']` I want it to fail, If I'll use any of these packages it will succeed

File: tests/conftest.py (hunk 19 above)
Comment (Hedingber) [reply tests/conftest.py:156]: added

File: mlrun/runtimes/base.py (hunk 6 above)
Comment (Hedingber) [reply mlrun/runtimes/base.py]: As I wrote above, the scenario is that the user's code finishs running, then the SDK code (in the pod) updates the **Run's state** to completed (here the Pod is still running), now, if an exception will be raised (because some bug in the wrapping SDK or something), the Pod state will be changed to failed (from running!), and the API monitoring the Pod's state, will change the Run's state from one terminal state - completed - to another - error

File: mlrun/api/api/utils.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/api/api/utils.py:128]: changed to run

File: tests/conftest.py (hunk 19 above)
Comment (Hedingber) [reply tests/conftest.py]: Talked offline, this whole thing isn't needed, I should have simply used the k8s client classes instead of tranforming dict to mock objects

File: mlrun/runtimes/base.py (hunk 2 above)
Comment (Hedingber) [reply mlrun/runtimes/base.py]: talked offline, leaving as is

File: mlrun/runtimes/base.py (hunk 6 above)
Comment (Hedingber) [reply mlrun/runtimes/base.py]: Talked offline, when the SDK will have an error we don't want to hide that so we're ok with the Run's state going from completed to failed.
We do identified that a race between SDK and server might occur so the Run's state will jitter: completed -> running -> completed, in the case where the API is monitoring the pod which is still running (cause SDK didn't finish yet) but the SDK already updated the Run's state to completed 
Will add a debounce logic for that

**Review by omesser — APPROVED** [review]
